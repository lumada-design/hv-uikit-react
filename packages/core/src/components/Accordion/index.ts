export { default as accordionClasses } from "./accordionClasses";
export * from "./accordionClasses";
export * from "./Accordion";
