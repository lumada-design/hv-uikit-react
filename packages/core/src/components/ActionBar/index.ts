export { default as actionBarClasses } from "./actionBarClasses";
export * from "./actionBarClasses";
export * from "./ActionBar";
