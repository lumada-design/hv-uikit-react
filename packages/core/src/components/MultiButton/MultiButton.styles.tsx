import clsx from "clsx";
import styled from "@emotion/styled";
import { theme } from "@hitachivantara/uikit-styles";
import { CSSProperties } from "react";
import { transientOptions } from "utils/transientOptions";

export const StyledRoot = styled(
  "div",
  transientOptions
)(({ $vertical }: { $vertical: boolean }) => ({
  display: "flex",
  height: 32,
  alignItems: "center",
  transition: "none",
  background: theme.colors.atmo2,
  position: "relative",
  zIndex: 0,

  // prevent the focus ring to be hidden by sibling hover background
  "&>.HvIsFocusVisible": {
    zIndex: 5,
  },

  ...($vertical && {
    flexDirection: "column",
    height: "auto",
    btnSecondary: {
      flex: "1 1 20px",
    },
    "& button.button": {
      width: "100%",
      borderLeft: `solid 1px ${theme.colors.atmo4}`,
      borderRight: `solid 1px ${theme.colors.atmo4}`,
      borderTop: "solid 1px transparent",
      borderBottom: "solid 1px transparent",
      "&:disabled": {
        color: theme.colors.atmo5,
        borderLeft: `solid 1px ${theme.colors.atmo4}`,
        borderRight: `solid 1px ${theme.colors.atmo4}`,
        borderTop: "solid 1px transparent",
        borderBottom: "solid 1px transparent",
        "&:hover": {
          borderLeft: `solid 1px ${theme.colors.atmo4}`,
          borderRight: `solid 1px ${theme.colors.atmo4}`,
          borderTop: "solid 1px transparent",
          borderBottom: "solid 1px transparent",
        },
      },
      "&:first-child": {
        borderTop: `solid 1px ${theme.colors.atmo4}`,
        borderTopLeftRadius: "2px",
        borderTopRightRadius: "2px",
      },
      "&:last-child": {
        borderBottom: `solid 1px ${theme.colors.atmo4}`,
        borderBottomLeftRadius: "2px",
        borderBottomRightRadius: "2px",
        "&:disabled:hover": {
          borderBottom: `solid 1px ${theme.colors.atmo4} !important`,
        },
      },
      "&:not(:first-child)": {
        marginLeft: 0,
        marginTop: -1,
      },
      "&.selected": {
        height: 32,
        width: `calc(100% + 2px) !important`,
        background: theme.colors.atmo1,
        ...(theme.typography.label as CSSProperties),
        borderRadius: "2px",
        border: `solid 1px ${theme.colors.acce1}`,
        zIndex: 2,
        "&:hover, &:focus": {
          background: theme.colors.atmo3,
        },
        "&:disabled": {
          zIndex: 1,
          color: theme.colors.atmo5,
          background: theme.colors.atmo1,
          border: `solid 1px ${theme.colors.atmo4}`,
        },
      },
    },
  }),
}));

export const StyledButton = (Element) =>
  styled(({ className }) => (
    <Element.type
      {...Element.props}
      className={clsx(
        "button",
        Element.props.selected && "selected",
        className
      )}
    />
  ))({
    height: 32,
    width: "100%",
    minWidth: 32,
    maxWidth: 200,
    padding: 0,
    transition: "none",
    flex: "1 0 0px",
    borderTop: `solid 1px ${theme.colors.atmo4}`,
    borderBottom: `solid 1px ${theme.colors.atmo4}`,
    borderLeft: "solid 1px transparent",
    borderRight: "solid 1px transparent",
    borderRadius: 0,
    ...(theme.typography.body as CSSProperties),
    "&:active": {
      backgroundColor: `${theme.colors.atmo3}`,
    },
    "&:disabled": {
      color: theme.colors.atmo5,
      borderTop: `solid 1px ${theme.colors.atmo4}`,
      borderBottom: `solid 1px ${theme.colors.atmo4}`,
      "&:hover": {
        borderTop: `solid 1px ${theme.colors.atmo4}`,
        borderBottom: `solid 1px ${theme.colors.atmo4}`,
        borderLeft: "solid 1px transparent",
        borderRight: "solid 1px transparent",
      },
    },
    "&:first-child": {
      borderLeft: `solid 1px ${theme.colors.atmo4}`,
      borderTopLeftRadius: "2px",
      borderBottomLeftRadius: "2px",
    },
    "&:last-child": {
      borderRight: `solid 1px ${theme.colors.atmo4}`,
      borderTopRightRadius: "2px",
      borderBottomRightRadius: "2px",
      "&:disabled:hover": {
        borderRight: `solid 1px ${theme.colors.atmo4} !important`,
      },
    },
    "&:not(:first-child)": {
      marginLeft: "-1px",
    },
    "&.selected": {
      background: theme.colors.atmo1,
      height: 34,
      ...(theme.typography.label as CSSProperties),
      borderRadius: "2px",
      border: `solid 1px ${theme.colors.acce1}`,
      zIndex: 2,
      "&:hover": {
        background: theme.colors.atmo3,
        "&:not(:disabled)": {
          border: `solid 1px ${theme.colors.acce1}`,
        },
        "&:disabled": {
          border: `solid 1px ${theme.colors.atmo4}`,
        },
      },
      "&:first-child, &:last-child": {
        border: `solid 1px ${theme.colors.acce1}`,
      },

      // prevent the focus ring to be hidden by sibling hover background
      // even when selected
      "&.HvIsFocusVisible": {
        zIndex: 5,
      },
      "&:disabled": {
        zIndex: 1,
        color: theme.colors.atmo5,
        background: theme.colors.atmo1,
        border: `solid 1px ${theme.colors.atmo4}`,
      },
    },
  });
