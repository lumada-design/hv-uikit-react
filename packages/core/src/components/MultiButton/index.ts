export { default as multiButtonClasses } from "./multiButtonClasses";
export * from "./multiButtonClasses";
export * from "./MultiButton";
