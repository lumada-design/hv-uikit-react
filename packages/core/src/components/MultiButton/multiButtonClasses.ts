import { getClasses } from "utils";

export type HvMultiButtonClasses = {
  root?: string;
  button?: string;
  vertical?: string;
  selected?: string;
};

const classKeys: string[] = ["root", "button", "vertical", "selected"];

const multiButtonClasses = getClasses<HvMultiButtonClasses>(
  classKeys,
  "HvMultiButton"
);

export default multiButtonClasses;
