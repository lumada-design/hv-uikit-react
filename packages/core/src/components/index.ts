export * from "./Box";
export { Box as HvBox } from "./Box";
export * from "./Button";
export { Button as HvButton } from "./Button";
export * from "./Dropdown";
export { Dropdown as HvDropdown } from "./Dropdown";
export * from "./Header";
export {
  Actions as HvHeaderActions,
  Brand as HvHeaderBrand,
  Header as HvHeader,
  Navigation as HvHeaderNavigation,
} from "./Header";
export * from "./Typography";
export { Typography as HvTypography } from "./Typography";
