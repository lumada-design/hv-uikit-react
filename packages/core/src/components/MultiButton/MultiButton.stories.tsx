import { useState } from "react";
import { range } from "lodash";
import { Meta, StoryObj } from "@storybook/react";
import { LocationPin, Map } from "@hitachivantara/uikit-icons";

import { HvButton, HvMultiButton, HvMultiButtonProps } from "components";

const meta: Meta<typeof HvMultiButton> = {
  title: "Inputs/MultiButton",
  component: HvMultiButton,
};
export default meta;

export const Main: StoryObj<HvMultiButtonProps> = {
  args: {
    disabled: false,
    vertical: false,
    variant: "secondarySubtle",
  },
  argTypes: {
    classes: { control: { disable: true } },
  },
  render: ({ vertical, disabled, variant }) => {
    {
      const [val, setVal] = useState(-1);

      return (
        <HvMultiButton
          style={{ width: "210px" }}
          vertical={vertical}
          disabled={disabled}
          variant={variant}
        >
          <HvButton
            key="1"
            startIcon={<Map />}
            selected={val === 0}
            onClick={() => setVal(0)}
          >
            Map
          </HvButton>
          <HvButton
            key="2"
            selected={val === 1}
            onClick={() => setVal(1)}
            startIcon={<LocationPin />}
          >
            Satellite
          </HvButton>
        </HvMultiButton>
      );
    }
  },
};

export const OnlyLabels: StoryObj<HvMultiButtonProps> = {
  render: () => {
    const [selection, setSelection] = useState(0);
    const buttons = ["Map", "Satellite"];

    const handleChange = (event, idx) => {
      setSelection(idx);
    };

    return (
      <HvMultiButton style={{ width: "210px" }}>
        {buttons.map((button, i) => (
          <HvButton
            id={button.toLowerCase()}
            key={`${buttons[i]}`}
            selected={selection === i}
            onClick={(evt) => handleChange(evt, i)}
          >
            {button}
          </HvButton>
        ))}
      </HvMultiButton>
    );
  },
};

export const OnlyIcons: StoryObj<HvMultiButtonProps> = {
  render: () => {
    const [selection, setSelection] = useState([0]);
    const buttons = [
      { name: "Map", icon: <Map /> },
      { name: "Location", icon: <LocationPin /> },
    ];

    const handleChange = (event, idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];
      setSelection(newSelection);
    };

    return (
      <HvMultiButton style={{ width: "64px" }}>
        {buttons.map(({ name, icon }, i) => (
          <HvButton
            id={name.toLowerCase()}
            key={`${buttons[i].name}`}
            icon
            aria-label={name}
            selected={selection.includes(i)}
            onClick={(evt) => handleChange(evt, i)}
          >
            {icon}
          </HvButton>
        ))}
      </HvMultiButton>
    );
  },
};

export const Disabled: StoryObj<HvMultiButtonProps> = {
  render: () => {
    const [selection, setSelection] = useState([0]);

    const toggleIndex = (idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];

      setSelection(newSelection);
    };

    return (
      <HvMultiButton disabled style={{ width: "320px" }}>
        <HvButton
          selected={selection.includes(0)}
          onClick={() => toggleIndex(0)}
        >
          Potato
        </HvButton>
        <HvButton
          selected={selection.includes(1)}
          onClick={() => toggleIndex(1)}
        >
          Onion
        </HvButton>
        <HvButton
          selected={selection.includes(2)}
          onClick={() => toggleIndex(2)}
        >
          Carrot
        </HvButton>
        <HvButton
          selected={selection.includes(3)}
          onClick={() => toggleIndex(3)}
        >
          Tomato
        </HvButton>
      </HvMultiButton>
    );
  },
};

export const DisabledItem: StoryObj<HvMultiButtonProps> = {
  render: () => {
    const [selection, setSelection] = useState([0]);

    const toggleIndex = (idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];

      setSelection(newSelection);
    };

    return (
      <HvMultiButton style={{ width: "320px" }}>
        <HvButton
          selected={selection.includes(0)}
          onClick={() => toggleIndex(0)}
        >
          Potato
        </HvButton>
        <HvButton
          selected={selection.includes(1)}
          onClick={() => toggleIndex(1)}
        >
          Onion
        </HvButton>
        <HvButton
          disabled
          selected={selection.includes(2)}
          onClick={() => toggleIndex(2)}
        >
          Carrot
        </HvButton>
        <HvButton
          selected={selection.includes(3)}
          onClick={() => toggleIndex(3)}
        >
          Tomato
        </HvButton>
      </HvMultiButton>
    );
  },
};

export const MultipleSelection: StoryObj<HvMultiButtonProps> = {
  render: () => {
    const [selection, setSelection] = useState([0, 2, 3, 5]);
    const buttons = [
      "Monday",
      "Tuesday",
      "Wednesday",
      "Thursday",
      "Friday",
      "Saturday",
      "Sunday",
    ];

    const handleChange = (event, idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];
      setSelection(newSelection);
    };

    return (
      <HvMultiButton style={{ width: "224px" }}>
        {buttons.map((button, i) => (
          <HvButton
            key={`${buttons[i]}`}
            aria-label={button}
            selected={selection.includes(i)}
            onClick={(evt) => handleChange(evt, i)}
          >
            {button[0]}
          </HvButton>
        ))}
      </HvMultiButton>
    );
  },
};

export const EnforcedSelection: StoryObj<HvMultiButtonProps> = {
  parameters: {
    docs: {
      description: {
        story:
          'HvMultiButton element set as enforced cannot be toggled - In this case **"Location 1" cannot be toggled on/off**.',
      },
    },
  },
  render: () => {
    const [selection, setSelection] = useState([0]);

    const handleChange = (event, idx) => {
      if (idx === 0) return; // enforced
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];
      setSelection(newSelection);
    };

    return (
      <div style={{ width: "600px" }}>
        <HvMultiButton>
          {range(5).map((i) => (
            <HvButton
              id={`location${i + 1 || ""}`}
              key={`ef-${i}`}
              startIcon={<LocationPin />}
              selected={selection.includes(i)}
              onClick={(evt) => handleChange(evt, i)}
            >
              {`Location ${i + 1}`}
            </HvButton>
          ))}
        </HvMultiButton>
      </div>
    );
  },
};

export const MinimumSelection: StoryObj<HvMultiButtonProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Specify a number of minimum elements that must be active - in this case a **minimum of 2**.",
      },
    },
  },
  render: () => {
    const [selection, setSelection] = useState([1, 2]);

    const handleChange = (event, idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];

      if (newSelection.length >= 2) setSelection(newSelection);
    };

    return (
      <div style={{ width: "800px" }}>
        <HvMultiButton>
          {range(5).map((i) => (
            <HvButton
              id={`location${i + 1}`}
              key={`ms-${i}`}
              startIcon={<LocationPin />}
              selected={selection.includes(i)}
              onClick={(evt) => handleChange(evt, i)}
            >
              {`Location ${i + 1}`}
            </HvButton>
          ))}
        </HvMultiButton>
      </div>
    );
  },
};

export const MaximumSelection: StoryObj<HvMultiButtonProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Specify a number of maximum elements that can be selected - in this case a **maximum of 2**.",
      },
    },
  },
  render: () => {
    const [selection, setSelection] = useState<number[]>([]);

    const handleChange = (event, idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];

      if (newSelection.length <= 2) setSelection(newSelection);
    };

    return (
      <div style={{ width: "800px" }}>
        <HvMultiButton>
          {range(5).map((i) => (
            <HvButton
              id={`location${i + 1}`}
              key={`maxse-${i}`}
              startIcon={<LocationPin />}
              selected={selection.includes(i)}
              onClick={(evt) => handleChange(evt, i)}
            >
              {`Location ${i + 1}`}
            </HvButton>
          ))}
        </HvMultiButton>
      </div>
    );
  },
};

export const VerticalOrientation: StoryObj<HvMultiButtonProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "MultiButton combinations with **vertical orientation** and **multiple selection**.",
      },
    },
  },
  render: () => {
    const [selection, setSelection] = useState([0, 2, 3, 5]);

    const buttons = [
      { name: "Map", icon: <Map />, key: 1 },
      { name: "Location", icon: <LocationPin />, key: 2 },
      { name: "Map", icon: <Map />, key: 3 },
      { name: "Location", icon: <LocationPin />, key: 4 },
      { name: "Map", icon: <Map />, key: 5 },
      { name: "Location", icon: <LocationPin />, key: 6 },
    ];

    const handleChange = (event, idx) => {
      const newSelection = selection.includes(idx)
        ? selection.filter((v) => v !== idx)
        : [...selection, idx];
      setSelection(newSelection);
    };

    return (
      <div style={{ display: "flex" }}>
        <HvMultiButton vertical style={{ width: "32px" }}>
          {buttons.map(({ name, icon }, i) => (
            <HvButton
              key={`${buttons[i].key}`}
              aria-label={name}
              selected={selection.includes(i)}
              onClick={(evt) => handleChange(evt, i)}
            >
              {icon}
            </HvButton>
          ))}
        </HvMultiButton>
        <HvMultiButton vertical style={{ marginLeft: "20px", width: "120px" }}>
          {buttons.map(({ name, icon }, i) => (
            <HvButton
              key={`${buttons[i].key}`}
              aria-label={name}
              startIcon={icon}
              selected={selection.includes(i)}
              onClick={(evt) => handleChange(evt, i)}
            >
              {name}
            </HvButton>
          ))}
        </HvMultiButton>
      </div>
    );
  },
};
