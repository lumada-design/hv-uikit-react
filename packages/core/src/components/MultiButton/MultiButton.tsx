import clsx from "clsx";
import React, { cloneElement } from "react";
import { HvButtonVariant } from "components";
import { HvBaseProps } from "types";
import { StyledButton, StyledRoot } from "./MultiButton.styles";

export type HvMultiButtonProps = HvBaseProps & {
  /** If all the buttons are disabled. */
  disabled?: boolean;
  /** If the MultiButton is to be displayed vertically. */
  vertical?: boolean;
  /** Category of button to use */
  variant?: HvButtonVariant;
  /** A Jss Object used to override or extend the styles applied to the empty state component. */
  classes?: {
    root?: string;
    button?: string;
  };
};

export const HvMultiButton = ({
  className,
  children,
  classes,
  disabled = false,
  vertical = false,
  variant = "secondarySubtle",
  ...others
}: HvMultiButtonProps) => {
  return (
    <StyledRoot
      className={clsx(className, classes?.root)}
      vertical={vertical}
      {...others}
    >
      {React.Children.map(children, (child) => {
        if (React.isValidElement(child)) {
          const childIsSelected = !!child.props.selected;
          const btn = cloneElement(child as React.ReactElement, {
            variant,
            disabled: disabled || child.props.disabled,
            "aria-pressed": childIsSelected,
          });
          const StyledBtn = StyledButton(btn);
          return React.createElement(StyledBtn);
        }
      })}
    </StyledRoot>
  );
};
