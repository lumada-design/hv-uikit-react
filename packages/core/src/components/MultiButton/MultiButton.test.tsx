import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";
import { LocationPin, Map } from "@hitachivantara/uikit-react-icons";
import { HvButton } from "components";
import { HvMultiButton } from "./MultiButton";

describe("MultiButton", () => {
  it("should be defined", () => {
    const { container } = render(<HvMultiButton />);
    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<HvMultiButton />);
    expect(container).toMatchSnapshot();
  });

  it("should render the buttons", () => {
    const { queryAllByRole } = render(
      <HvMultiButton>
        <HvButton key="1" startIcon={<Map />}>
          Button1
        </HvButton>
        <HvButton key="2" startIcon={<LocationPin />}>
          Button2
        </HvButton>
      </HvMultiButton>
    );
    const buttons = queryAllByRole("button");
    expect(buttons.length).toBe(2);
  });
});
