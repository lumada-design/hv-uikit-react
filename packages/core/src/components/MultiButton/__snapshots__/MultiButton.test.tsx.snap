// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiButton > should render correctly 1`] = `
<div>
  <div
    class="HvMultiButton-root css-fxit71-StyledRoot e17vrbb41"
  />
</div>
`;
